src/Components/Web/src/Virtualization/Virtualize.cs

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components.Rendering;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.Virtualization
+{
+    /// <summary>
+    /// Provides functionality for rendering a virtualized, fixed-size list of items.
+    /// </summary>
+    /// <typeparam name="TItem">The <c>context</c> type for the items being rendered.</typeparam>
+    public sealed class Virtualize<TItem> : VirtualizeBase<TItem>, IAsyncDisposable
+    {
+        private const string JsFunctionsPrefix = "Blazor._internal.Virtualize";
+
+        private DotNetObjectReference<Virtualize<TItem>>? _selfReference;
+
+        private int _itemsBefore;
+
+        private int _itemsVisible;
+
+        private ElementReference _spacerBefore;
+
+        private ElementReference _spacerAfter;
+
+        [Inject]
+        private IJSRuntime JSRuntime { get; set; } = default!;
+
+        /// <inheritdoc />
+        protected override async Task OnAfterRenderAsync(bool firstRender)
+        {
+            if (firstRender)
+            {
+                _selfReference = DotNetObjectReference.Create<Virtualize<TItem>>(this);
+                await JSRuntime.InvokeVoidAsync($"{JsFunctionsPrefix}.init", _selfReference, _spacerBefore, _spacerAfter);
+            }
+        }
+
+        /// <inheritdoc />
+        protected override void BuildRenderTree(RenderTreeBuilder builder)
+        {
+            var itemsAfter = Math.Max(0, ItemSource.Count() - _itemsVisible - _itemsBefore);
+
+            builder.OpenElement(0, "div");
+            builder.AddAttribute(1, "style", GetSpacerStyle(_itemsBefore));
+            builder.AddElementReferenceCapture(2, elementReference => _spacerBefore = elementReference);
+            builder.CloseElement();
+
+            builder.OpenRegion(3);
+
+            if (ItemTemplate != null)
+            {
+                var items = IsVirtualized ? ItemSource.Skip(_itemsBefore).Take(_itemsVisible) : ItemSource;
+
+                foreach (var item in items)
+                {
+                    ItemTemplate(item)(builder);
+                }
+            }
+
+            builder.CloseRegion();
+
+            builder.OpenElement(4, "div");
+            builder.AddAttribute(5, "style", GetSpacerStyle(itemsAfter));
+            builder.AddElementReferenceCapture(6, elementReference => _spacerAfter = elementReference);
+
+            builder.CloseElement();
+
+            if (Footer != null)
+            {
+                builder.OpenRegion(7);
+
+                Footer.Invoke(builder);
+
+                builder.CloseRegion();
```

Comment:
The builder has built-in support for adding renderfragments, so I'd suggest:

```suggestion
                builder.AddContent(7, Footer);
```

Then you don't need the `if` around it either, because it's a no-op if the renderfragment is null.